src/coreclr/src/vm/i386/excepx86.cpp

```diff
@@ -1868,8 +1868,10 @@ NOINLINE LPVOID COMPlusEndCatchWorker(Thread * pThread)
     // Sync managed exception state, for the managed thread, based upon any active exception tracker
     pThread->SyncManagedExceptionState(fIsDebuggerHelperThread);
 
-    if (InlinedCallFrame::FrameHasActiveCall(pThread->m_pFrame))
+    if ((pThread->m_pFrame < esp) && InlinedCallFrame::FrameHasActiveCall(pThread->m_pFrame))
```

Comment:
How does this work?

I believe that we need to check whether the frame is within the method that we are resuming into, ie we need to check whether `pFrame < callerESP`.

Reply:
Hmm, you are right that this would not work for the case when the location we resume to is in the same function as the pinvoke frame. It works when it is anywhere up the call chain.

Reply:
I would not expect the current version of the change to fix the problem.

Anyway, this is not a new problem. #1929 was just unlucky to hit it. I have opened #2240 to track it and added workaround to #1929 to make forward progress.